ignition_ros2_control/src/ignition_ros2_control_plugin.cpp

```diff
@@ -0,0 +1,477 @@
+    return;
+  }
+
+  auto cm_update_rate = this->dataPtr->controller_manager_->get_parameter("update_rate").as_int();
+  this->dataPtr->control_period_ = rclcpp::Duration(
+    std::chrono::duration_cast<std::chrono::nanoseconds>(
+      std::chrono::duration<double>(1.0 / static_cast<double>(cm_update_rate))));
+
+  this->dataPtr->entity_ = _entity;
+}
+
+//////////////////////////////////////////////////
+void IgnitionROS2ControlPlugin::PreUpdate(
+  const ignition::gazebo::UpdateInfo & _info,
+  ignition::gazebo::EntityComponentManager & /*_ecm*/)
+{
+  static bool warned{false};
+  if (!warned) {
+    rclcpp::Duration gazebo_period(_info.dt);
+
+    // Check the period against the simulation period
+    if (this->dataPtr->control_period_ < _info.dt) {
+      RCLCPP_ERROR_STREAM(
+        this->dataPtr->node->get_logger(),
+        "Desired controller update period (" << this->dataPtr->control_period_.seconds() <<
+          " s) is faster than the gazebo simulation period (" <<
+          gazebo_period.seconds() << " s).");
+    } else if (this->dataPtr->control_period_ > gazebo_period) {
+      RCLCPP_WARN_STREAM(
+        this->dataPtr->node->get_logger(),
+        " Desired controller update period (" << this->dataPtr->control_period_.seconds() <<
+          " s) is slower than the gazebo simulation period (" <<
+          gazebo_period.seconds() << " s).");
+    }
+    warned = true;
+  }
+
+  // Always set commands on joints, otherwise at low control frequencies the joints tremble
+  // as they are updated at a fraction of gazebo sim time
+  this->dataPtr->controller_manager_->write();
+}
+
+//////////////////////////////////////////////////
+void IgnitionROS2ControlPlugin::PostUpdate(
+  const ignition::gazebo::UpdateInfo & _info,
+  const ignition::gazebo::EntityComponentManager & /*_ecm*/)
+{
+  // Get the simulation time and period
+  rclcpp::Time sim_time_ros(std::chrono::duration_cast<std::chrono::nanoseconds>(
+      _info.simTime).count(), RCL_ROS_TIME);
+  rclcpp::Duration sim_period = sim_time_ros - this->dataPtr->last_update_sim_time_ros_;
+
+  if (sim_period >= this->dataPtr->control_period_) {
+    this->dataPtr->last_update_sim_time_ros_ = sim_time_ros;
+    auto ign_controller_manager =
+      std::dynamic_pointer_cast<ignition_ros2_control::IgnitionSystemInterface>(
+      this->dataPtr->controller_manager_);
+    this->dataPtr->controller_manager_->read();
+    this->dataPtr->controller_manager_->update();
+  }
+}
+}  // namespace ignition_ros2_control
+
+IGNITION_ADD_PLUGIN(
+  ignition_ros2_control::IgnitionROS2ControlPlugin,
+  ignition::gazebo::System,
+  ignition_ros2_control::IgnitionROS2ControlPlugin::ISystemConfigure,
+  ignition_ros2_control::IgnitionROS2ControlPlugin::ISystemPreUpdate,
+  ignition_ros2_control::IgnitionROS2ControlPlugin::ISystemPostUpdate)
+
+IGNITION_ADD_PLUGIN_ALIAS(
+  ignition_ros2_control::IgnitionROS2ControlPlugin,
+  "ignition_ros2_control::IgnitionROS2ControlPlugin")
```

Comment:
Is the alias exactly the same as the original name?

On `ign-gazebo`, we set aliases to remove the hidden inline namespace (see the `v6` for example):

```
$ ign plugin -i -v -p install/lib/ign-gazebo-6/plugins/libignition-gazebo6-diff-drive-system.so
Loading plugin library file [install/lib/ign-gazebo-6/plugins/libignition-gazebo6-diff-drive-system.so]
Loader State
        Known Interfaces: 4
                ignition::gazebo::v6::System
                ignition::gazebo::v6::ISystemPreUpdate
                ignition::gazebo::v6::ISystemPostUpdate
                ignition::gazebo::v6::ISystemConfigure
        Known Plugins: 1
                [ignition::gazebo::v6::systems::DiffDrive]
                        has 1 alias:
                                [ignition::gazebo::systems::DiffDrive]
                        implements 4 interfaces:
                                ignition::gazebo::v6::ISystemConfigure
                                ignition::gazebo::v6::ISystemPostUpdate
                                ignition::gazebo::v6::ISystemPreUpdate
                                ignition::gazebo::v6::System
```

But here I think there's no need to add an alias